runtime/src/utils/buffer/pool.rs

```diff
@@ -0,0 +1,443 @@
+use crate::{buffer::Buffer, Blob, Error, RwLock};
+use commonware_utils::StableBuf;
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+
+/// A [BufferPool] caches pages of [Blob] data in memory.
+///
+/// A single buffer pool can be used to cache data from multiple blobs by assigning a unique id to
+/// each.
+pub struct BufferPool<const PAGE_SIZE: usize> {
+    /// The page cache index, indexed by the blob id and the page number, mapping to the index of
+    /// the cache entry for the page.
+    index: HashMap<(u32, u64), usize>,
+
+    /// The page cache.
+    cache: Vec<CacheEntry<PAGE_SIZE>>,
+
+    /// The current clock hand index into `cache` for the CLOCK replacement policy.
+    clock: usize,
+
+    /// The next id to assign to a blob that will be managed by this pool.
+    next_id: u32,
+
+    /// The maximum number of pages that will be cached.
+    capacity: usize,
+}
+
+struct CacheEntry<const PAGE_SIZE: usize> {
+    /// The cache key which is composed of the blob id and page number of the page.
+    key: (u32, u64),
+
+    /// A bit indicating whether this page was recently referenced.
+    referenced: AtomicBool,
+
+    /// The cached page itself.
+    data: Box<[u8; PAGE_SIZE]>,
```

Comment:
Given this is already heap allocated, I wonder if we actually benefit from using a const `PAGE_SIZE` here?

If we dropped it (and made it a const), it may make using this thing a tad simpler?

Reply:
I suppose it means we don't need to store the "len" and "capacity" like you would with a Vec.

Reply:
Right, maybe not all that significant, but with enough pages in the cache it adds up.